llvm/include/llvm/CodeGen/GlobalISel/Utils.h

```diff
@@ -593,5 +593,31 @@ bool isGuaranteedNotToBeUndef(Register Reg, const MachineRegisterInfo &MRI,
 /// estimate of the type.
 Type *getTypeForLLT(LLT Ty, LLVMContext &C);
 
+enum class GIConstantKind { Scalar, FixedVector, ScalableVector };
+
+/// An integer-like constant.
+class GIConstant {
+  GIConstantKind Kind;
+  SmallVector<APInt> Values;
+  APInt Value;
+
+public:
+  GIConstant(ArrayRef<APInt> Values)
+      : Kind(GIConstantKind::FixedVector), Values(Values) {};
+  GIConstant(const APInt &Value, GIConstantKind Kind)
+      : Kind(Kind), Value(Value) {};
+
+  GIConstantKind getKind() const { return Kind; }
+
+  APInt getScalarValue() const;
+
+  static std::optional<GIConstant> getConstant(Register Const,
+                                               const MachineRegisterInfo &MRI);
+};
+
+/// Return true if the given value is known to be non-zero when defined.
+bool isKnownNonZero(Register Reg, const MachineRegisterInfo &MRI,
+                    GISelKnownBits *KB, unsigned Depth = 0);
+
```

Comment:
We have similar interfaces like `isConstantOrConstantVector` and friends. Can we extend those or unify in some more consistent way rather than have 2 different ways of querying constant values?

Reply:
Note that `isConstantOrConstantVector`starts with `is`. It is a test. `GIConstant::getConstant` gives you a constant. It is a vector of `APInt` s. We can add arbitrary arithmetic. `GIConstant + GIConstant`.

Reply:
Yes I'm aware, I said `isConstantOrConstantVector` and *friends*. One of those friends is this:
```
std::optional<APInt>
llvm::isConstantOrConstantSplatVector(
```
which returns an APInt. So if we add GIConstant now we have even more ways to do similar things. I'm not saying your approach is necessarily wrong here, but taking a step back maybe this change should be a separate effort to unify our utilities better.

Reply:
We were limit to splat vectors before. `GIConstant` does not have this restriction.

Reply:
I am fine with a family of tests for various kinds of constnesses, but I would argue that `GIConstant` is powerful enough to become **the** constant. It is a class. We can add arbitrary member functions for tests and arithmetic. Constants can be scalar, fixed-, and scalable-vectors.

Reply:
The design is minimal, it has uses, and getting rid of the splat vector variants will take a lot of energy. I would prefer to continue with this PR.

In a separate change, it would be dead code.